services/vision/obstacledepth/obstacle_depth.go

```diff
@@ -0,0 +1,370 @@
+// Package obstacledepth uses an underlying depth camera to fulfill GetObjectPointClouds,
+// using the method outlined in (Manduchi, Roberto, et al. "Obstacle detection and terrain classification
+// for autonomous off-road navigation." Autonomous robots 18 (2005): 81-102.)
+package obstacledepth
+
+import (
+	"context"
+	"image"
+	"math"
+	"strconv"
+	"sync"
+
+	"github.com/edaniels/golog"
+	"github.com/golang/geo/r3"
+	"github.com/muesli/clusters"
+	"github.com/muesli/kmeans"
+	"github.com/pkg/errors"
+	"go.opencensus.io/trace"
+
+	"go.viam.com/rdk/components/camera"
+	"go.viam.com/rdk/pointcloud"
+	"go.viam.com/rdk/resource"
+	"go.viam.com/rdk/rimage"
+	"go.viam.com/rdk/rimage/transform"
+	"go.viam.com/rdk/robot"
+	svision "go.viam.com/rdk/services/vision"
+	"go.viam.com/rdk/spatialmath"
+	"go.viam.com/rdk/utils"
+	vision "go.viam.com/rdk/vision"
+	"go.viam.com/rdk/vision/segmentation"
+)
+
+var model = resource.DefaultModelFamily.WithModel("obstacle_depth")
+
+// ObstaclesDepthConfig specifies the parameters to be used for the obstacle depth service.
+type ObstaclesDepthConfig struct {
+	K          int     `json:"k"`
+	Hmin       float64 `json:"hmin"`
+	Hmax       float64 `json:"hmax"`
+	ThetaMax   float64 `json:"theta_max"`
+	ReturnPCDs bool    `json:"return_pcds"`
+	intrinsics *transform.PinholeCameraIntrinsics
+}
+
+// obsPoint is a useful intermediate struct (which pts are obstacles).
+type obsPoint struct {
+	point      image.Point
+	isObstacle bool
+	depth      rimage.Depth
+}
+
+// obsDepth is the underlying struct actually used by the service.
+type obsDepth struct {
+	dm         *rimage.DepthMap
+	ptChunks   [][]obsPoint
+	ptList     []obsPoint
+	hMin       float64
+	hMax       float64
+	sinTheta   float64
+	intrinsics *transform.PinholeCameraIntrinsics
+	returnPCDs bool
+	k          int
+}
+
+const (
+	// the first 3 consts are parameters from Manduchi et al.
+	defaultHmin     = 0.0
+	defaultHmax     = 150.0
+	defaultThetamax = math.Pi / 4
+	defaultK        = 10 // default number of obstacle segments to create
+	// the last 2 consts are hyperparameters that can be tweaked for performance improvement.
+	chunkSize = 200 // we send chunkSize points in each goroutine
+	sampleN   = 4   // we sample 1 in every sampleN depth points
+)
+
+func init() {
+	resource.RegisterService(svision.API, model, resource.Registration[svision.Service, *ObstaclesDepthConfig]{
+		DeprecatedRobotConstructor: func(ctx context.Context, r any, c resource.Config, logger golog.Logger) (svision.Service, error) {
+			attrs, err := resource.NativeConfig[*ObstaclesDepthConfig](c)
+			if err != nil {
+				return nil, err
+			}
+			actualR, err := utils.AssertType[robot.Robot](r)
+			if err != nil {
+				return nil, err
+			}
+			return registerObstacleDepth(ctx, c.ResourceName(), attrs, actualR)
+		},
+	})
+}
+
+// Validate ensures all parts of the config are valid.
+func (config *ObstaclesDepthConfig) Validate(path string) ([]string, error) {
```

Comment:
Are there any other parameters we need to check are non-negative? (Hmin, Hmax, ThetaMax)